flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/SplitAggregateRule.scala

```diff
@@ -280,11 +282,15 @@ class SplitAggregateRule extends RelOptRule(
     }
 
     // STEP 2.3: construct partial aggregates
-    relBuilder.aggregate(
-      relBuilder.groupKey(fullGroupSet, ImmutableList.of[ImmutableBitSet](fullGroupSet)),
+    // Create aggregate node directly to avoid ClassCastException,
+    // Please see https://issues.apache.org/jira/browse/FLINK-21923 for more details.
+    val partialAggregate = FlinkLogicalAggregate.create(
```

Comment:
The current approach does not consider the optimization of aggregate function reuse

Reply:
The implementation of the previous version used calcite's function reuse optimization.
It seems that only avg will translate count and sum0. Discuss offline with @beyond1920, use the create method to be more direct, otherwise, we need to add a project to restore the columns.
Do you suggest to still use calcite, and then rebuild the project?

Reply:
For a quick fix, I suggest we keep the current approach and create a new issue to tract the optimization (please also add some comments here).  you can take the optimization issue if you like

Reply:
Thanks for your reminder! I has created a new issue [FLINK-22412](https://issues.apache.org/jira/browse/FLINK-22412)

Reply:
can you add a TODO here (such as `// TODO reuse aggregate function, see FLINK-22412`)
 and add some tests about window aggregate

Reply:
ok，I will deal with it as soon as possible

Reply:
I have added TODO, please take a look again, thanks